### ISSUE TITLE
[ @typescript-eslint/ban-types] Banning of primitive object types also aplies when accessing the type of a property

### ISSUE THREAD

**Issue by KilianKilmister** [typescript-eslint/typescript-eslint#2548]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

rule:

```JS
{
  "rules": {
    "@typescript-eslint//ban-types": ["error", { // default settings
  String: {
    message: 'Use string instead',
    fixWith: 'string',
  },
  Boolean: {
    message: 'Use boolean instead',
    fixWith: 'boolean',
  },
  Number: {
    message: 'Use number instead',
    fixWith: 'number',
  },
  Symbol: {
    message: 'Use symbol instead',
    fixWith: 'symbol',
  },

  Function: {
    message: [
      'The `Function` type accepts any function-like value.',
      'It provides no type safety when calling the function, which can be a common source of bugs.',
      'It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.',
      'If you are expecting the function to accept certain arguments, you should explicitly define the function shape.',
    ].join('\n'),
  },

  // object typing
  Object: {
    message: [
      'The `Object` type actually means "any non-nullish value", so it is marginally better than `unknown`.',
      '- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.',
      '- If you want a type meaning "any value", you probably want `unknown` instead.',
    ].join('\n'),
  },
  '{}': {
    message: [
      '`{}` actually means "any non-nullish value".',
      '- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.',
      '- If you want a type meaning "any value", you probably want `unknown` instead.',
    ].join('\n'),
  },
  object: {
    message: [
      'The `object` type is currently hard to use ([see this issue](https://github.com/microsoft/TypeScript/issues/21732)).',
      'Consider using `Record<string, unknown>` instead, as it allows you to more easily inspect and use the keys.',
    ].join('\n'),
  },
};]
  }
}
```

example type

```TS
type ReplacerFunction = Parameters<String['replace']>[1]
```

`compilerOptions` in `tsconfig`

```json5
{
  "lib": [
    "esnext"
  ],
  "alwaysStrict": true,
  "stripInternal": true,
  "assumeChangesOnlyAffectDirectDependencies": true,
  "declaration": true,
  "declarationDir": "@types",
  "declarationMap": true,
  "diagnostics": true,
  "forceConsistentCasingInFileNames": false,
  "importsNotUsedAsValues": "error",
  "incremental": true,
  "module": "esnext",
  "moduleResolution": "node",
  "outDir": "dist",
  "plugins": [],
  "pretty": true,
  "rootDir": ".",
  "baseUrl": ".",
  "sourceMap": true,
  "target": "ESNext",
  "useDefineForClassFields": true,
  "experimentalDecorators": true,
  "strict": true,
  "strictBindCallApply": true,
  "strictFunctionTypes": true,
  "strictNullChecks": true,
  "noImplicitAny": true,
  "strictPropertyInitialization": true,
  "noImplicitReturns": false,
  "noImplicitThis": true,
  "allowSyntheticDefaultImports": true,
  "emitDecoratorMetadata": true,
  "composite": true,
  "preserveSymlinks": true
},
```

**Expected Result**

Ignore the use as it can't be replaced with a primitive type

**Actual Result**

```ts
type ReplacerFunction = Parameters<string['replace']>[1] // which will obviously not work
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `3.10.1` |
| `TypeScript`                       | `4.1.0-dev.20200911` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `14.10` |

**Comment by bradzacher** [comment]
The rule is intentionally simple. It simply looks for banned names and reports on them.

If you have an edge case where you have to use a banned type, use a disable comment.

**Comment by KilianKilmister** [comment]
@bradzacher i wouldn't call this an edge-case and it affects the default configuration of the rule. this is a really unsatisfying answer

**Comment by bradzacher** [comment]
Types are banned for all manner of reasons:  Some codebases prefer one type over another. Some codebases just don't want a certain type to be used.

The rule handles all of these cases in the same way - it report on a usage of the type.
Lint rules should never be treated as "do this thing or else" - lint rules are an automated way to enforce a coding practice.

If you have a case that violates the lint rule you can either: reconfigure the lint rule, or you can ignore the case with a disable comment.

> i wouldn't call this an edge-case

The rule has existed and been in the recommended set for almost 2 years.
You're the first one to raise an issue about this.
It sounds like an edge case to me?

**Comment by KilianKilmister** [comment]
> It sounds like an edge case to me?

I looked at it as a common practice, but that's not relevant.

There are similar issues around this rule though: #2121 #830

The documentation should at least be extended to mention that "type" isn't an actual typescript type but just a string match on type names
